central/graphql/resolvers/clusters.go

```diff
@@ -145,6 +149,53 @@ func (resolver *Resolver) Clusters(ctx context.Context, args PaginatedQuery) ([]
 	return resolver.wrapClustersWithContext(ctx, clusters, err)
 }
 
+func (resolver *Resolver) clustersForPermission(ctx context.Context, args PaginatedQuery, resource permissions.ResourceWithAccess) ([]*scopeObjectResolver, error) {
+	query, err := args.AsV1QueryOrEmpty()
+	if err != nil {
+		return nil, err
+	}
+	clusterIDs, unrestricted, err := sacHelper.ListClusterIDsInScope(ctx, []permissions.ResourceWithAccess{resource})
+	if err != nil {
+		return nil, err
+	}
+	var elevatedCtx context.Context
+
+	// Elevate context to find all matching clusters.
+	// The access control is enforced by the query restriction to
+	// the cluster IDs in the requester scope.
+	if unrestricted {
+		elevatedCtx = sac.WithGlobalAccessScopeChecker(
+			ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.Cluster),
+			),
+		)
+	} else {
+		elevatedCtx = sac.WithGlobalAccessScopeChecker(
+			ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.Cluster),
+				sac.ClusterScopeKeys(clusterIDs.AsSlice()...),
+			),
+		)
+	}
+	clusters, err := resolver.ClusterDataStore.SearchRawClusters(elevatedCtx, query)
```

Comment:
It would be good to cover with a test that a "hand-crafted" query will not allow access to clusters that are not part of the user's scope.